developer_guides/firmware/async_messaging_best_practices.rst

```diff
@@ -0,0 +1,651 @@
+     int32_t error;
+     queue_t *request_queue;
+   };
+
+   message_context_t message_context;
+
+   void callback(const ams_message_payload_s * const ams_message_payload, void *ctx) {
+      if(ams_message_payload->message_length != sizeof(request_t)) {
+         ((message_context_t *)ctx)->error = 1;
+         return;
+      }
+
+      queue_push(((message_context_t *)ctx)->request_queue, (request_t *)(ams_message_payload->message));
+   }
+
+   error_code componentA::Init(Pipeline* parent_ppl, const ModuleACfg* cfg) {
+     //getting Message Type ID
+     error = am_service_get_message_type_id(MESSAGE_A, &message_type_id);
+     if(error != 0) {
+       //error handling
+     }
+
+     //registering consumer
+     message_context.error = 0;
+     message_context.request_queue = &request_queue;
+     error = am_service_register_consumer(message_type_id, GetcomponentID(), GetInstanceID(), callback, &message_context);
+     if(error != 0) {
+       //error handling
+     }
+   }
+
+   error_code componentA::ProcessData(size_t max_output_data_size, uint32_t* custom_error_code) {
+      while(queue_size(message_context.request_queue) > 0) {
+         if(message_context.error != 0) {
+           //error handling
+         }
+
+         request_t *r = queue_pop(message_context.request_queue);
+         //handling of async message
+         process_request(r);
+      }
+
+      //regular processing
+   }
+
+.. _processing_in_a_consumer_callback:
+
+Processing in a consumer callback
+*********************************
+
+The consumer callbacks must not do any heavy processing. Best if the callback
+only saves information and a component will process it in the next processing
+slot.
+
+-  The asynchronous message callbacks are called in the context of a message
+   producer or in the context of IDC task.
+-  If a heavy processing is done in a message, then a producer MPCS or IDC task
+   budget needs to take into account this heavy processing. While it may be
+   possible to take into account additional budget for simple cases, it is
+   impossible to accommodate MCPS requirements in the general case (i.e. large
+   number of consumers with heavy processing).
+
+The example described in :ref:`message_handling` is also applicable for this issue.
+
+.. _message_filtering_mechanism:
+
+Message Filtering Mechanism
+***************************
+
+1. If a consumer is interested in async messages from one particular producer,
+   it needs to register for that particular producer using component ID and
+   instance ID. A component ID and instance ID of producer should be passed as
+   ``LARGE_CONFIG_SET`` parameter to the consumer.
+2. If a producer wants to target a specific component instance, then it should
+   use send function that is parametrized with component ID and instance ID.
+
+Passing Pointers In Asynchronous Messages
+*****************************************
+
+The asynchronous messages do not prevent to pass pointers between components. It
```

Comment:
If we can pass ptrs, who would own and set memory permissions for that pointer. I assume the sender would have to make sure the receiver has correct permissions and memory is valid etc.

Reply:
@lgirdwood the purpose of this paragraph is to discourage user/developers to pass the pointers in the first place because it can lead to multiple undefined behaviors unless synchronized correctly.